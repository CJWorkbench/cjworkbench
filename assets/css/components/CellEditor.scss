
.cell-edits--column {
	margin-bottom: 1rem;
}

.cell-edits--row {
	background-color: #fff;
	width:none;
	display: flex;
}

.cell-edits--row .text-field {
	white-space: normal;
	padding:0;
	line-height:2rem;
	width:17rem;
	padding: 0.7rem;

}

.cell-edits--row .number-field{
	display: flex;
	justify-content: center;
	align-items: flex-start;
	padding:0.7rem;
	line-height:2rem;
	width: 5rem;
	color:#b4b4b4;
}
