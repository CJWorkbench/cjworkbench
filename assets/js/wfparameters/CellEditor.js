// Displays edits made to individual cells

import React from 'react';
import PropTypes from 'prop-types'

export default class CellEditor extends React.Component {
  // Starting from a JSON string that represents an array like this
  // [
  //   { row: 3, col: 'foo', value:'bar' },
  //   { row: 6, col: 'food', value:'sandwich' },
  //   ...
  // ]
  // convert it to one that combines all edits for single column, like this
  // [
  //   { col: 'foo', edits: [ { row: 3, value: 'bar'}, ... ] },
  //   { col: 'food', edits: [ { row: 6, value: 'sandwich}, ... ] },
  // ]

  parseEditList(edit_json) {

    let unsorted_edits;
    try {
      unsorted_edits = JSON.parse(edit_json);
    } catch(err) {
      return [];    // error parsing JSON, most likely because of empty parameter
    }

    let edits = {};

    for (let edit of unsorted_edits) {
      let colName = edit.col;
      let colEdits = edits[colName] || [];
      colEdits.push({ row: edit['row'], value: edit['value']});
      edits[colName] = colEdits;
    }

    // Sort column names and convert to array
    let colNames = Object.keys(edits);
    colNames.sort();

    let editsArray = colNames.map( (colName) => {
      return { col: colName, edits: edits[colName] };
    });

    return editsArray;
  }


  render() {

    let edits = this.parseEditList(this.props.edits);

    // Each column gets a div
    var editList = edits.map((colEdits) => {

      // Each row in this list of edits for this column also gets div love
      let rows = colEdits['edits'].map( (row, idx) => {
        return (
          <div className='cell-edits--row' key={idx}>
            <div className='edited-row'>{row['row']+1}</div>
            <div className='edited-text'>{row['value']}</div>
          </div>
        );
      });

      let colName = colEdits['col'];
      return (
        <div className='cell-edits--column' key={colName}>
          <div className='edited-column'>{ colName }</div>
          <div className='cell-edits--rows'>
            { rows }
          </div>
        </div>
      );
    });

    return  (
      <div>
        { editList }
      </div>
    );
  }
}

CellEditor.propTypes = {
  edits:        PropTypes.string.isRequired,
  onSave:       PropTypes.func.isRequired,
};
