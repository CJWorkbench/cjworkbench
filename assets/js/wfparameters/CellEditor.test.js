import React from 'react'
import CellEditor from './CellEditor'
import { mount } from 'enzyme'

describe('CellEditor', () => {

  const testParam = '[\n' +
    '    { "row": 3, "col": "foo", "value":"bar" },\n' +
    '    { "row": 6, "col": "food", "value":"sandwich" },\n' +
    '    { "row": 5, "col": "foo", "value":"gak" },\n' +
    '    { "row": 17,  "col": "food", "value":"pizza" }\n' +
    ' ]';

  it('Renders correctly', () => {
    var wrapper = mount(
      <CellEditor
        edits={testParam}
        onSave={() => {}}
      />);

    expect(wrapper.find('.cell-edits--column')).toHaveLength(2);
    expect(wrapper.find('.cell-edits--row')).toHaveLength(4);

    expect(wrapper).toMatchSnapshot();
  });

  it('Renders empty data', () => {
     var wrapper = mount(
      <CellEditor
        edits={''}
        onSave={() => {}}
      />);

    expect(wrapper.find('.cell-edits--column')).toHaveLength(0);
    expect(wrapper.find('.cell-edits--row')).toHaveLength(0);

    expect(wrapper).toMatchSnapshot();
  });

});
