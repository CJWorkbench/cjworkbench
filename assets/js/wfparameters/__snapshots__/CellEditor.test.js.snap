// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellEditor Renders correctly 1`] = `
<CellEditor
  edits="[
    { \\"row\\": 3, \\"col\\": \\"foo\\", \\"value\\":\\"bar\\" },
    { \\"row\\": 6, \\"col\\": \\"food\\", \\"value\\":\\"sandwich\\" },
    { \\"row\\": 5, \\"col\\": \\"foo\\", \\"value\\":\\"gak\\" },
    { \\"row\\": 17,  \\"col\\": \\"food\\", \\"value\\":\\"pizza\\" }
 ]"
  onSave={[Function]}
>
  <div>
    <div
      className="cell-edits--column content-3 label-margin"
    >
      foo
      <div
        className="cell-edits--rows t-d-gray content-3 "
      >
        <div
          className="cell-edits--row"
        >
          <div
            className="number-field"
          >
            4
          </div>
          <div
            className="text-field"
          >
            bar
          </div>
        </div>
        <div
          className="cell-edits--row"
        >
          <div
            className="number-field"
          >
            6
          </div>
          <div
            className="text-field"
          >
            gak
          </div>
        </div>
      </div>
    </div>
    <div
      className="cell-edits--column content-3 label-margin"
    >
      food
      <div
        className="cell-edits--rows t-d-gray content-3 "
      >
        <div
          className="cell-edits--row"
        >
          <div
            className="number-field"
          >
            7
          </div>
          <div
            className="text-field"
          >
            sandwich
          </div>
        </div>
        <div
          className="cell-edits--row"
        >
          <div
            className="number-field"
          >
            18
          </div>
          <div
            className="text-field"
          >
            pizza
          </div>
        </div>
      </div>
    </div>
  </div>
</CellEditor>
`;

exports[`CellEditor Renders empty data 1`] = `
<CellEditor
  edits=""
  onSave={[Function]}
>
  <div />
</CellEditor>
`;
